src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
 
   _target_table = StorageManager::get().get_table(_target_table_name);
 
-  // These TypedSegmentProcessors kind of retrieve the template parameter of the segments.
-  auto typed_segment_processors = std::vector<std::unique_ptr<AbstractTypedSegmentProcessor>>();
-  for (const auto& column_type : _target_table->column_data_types()) {
-    typed_segment_processors.emplace_back(
-        make_unique_by_data_type<AbstractTypedSegmentProcessor, TypedSegmentProcessor>(column_type));
-  }
-
-  auto total_rows_to_insert = static_cast<uint32_t>(input_table_left()->row_count());
+  Assert(_target_table->max_chunk_size() > 0, "Expected max chunk size of target table to be greater than zero");
 
-  // First, allocate space for all the rows to insert. Do so while locking the table to prevent multiple threads
-  // modifying the table's size simultaneously.
-  auto start_index = 0u;
-  auto start_chunk_id = ChunkID{0};
-  auto end_chunk_id = 0u;
+  /**
+   * 1. Allocate the required rows in the target Table, without actually copying data to them.
+   *    Do so while locking the table to prevent multiple threads modifying the table's size simultaneously.
+   *    Since allocation is expected to be faster than writing to the memory, allocating under lock and then writing -
+   *    in a second step - without lock will minimize the time that the Table's append_mutex is locked.
+   */
   {
-    auto scoped_lock = _target_table->acquire_append_mutex();
+    const auto append_lock = _target_table->acquire_append_mutex();
+
+    auto remaining_rows = input_table_left()->row_count();
 
     if (_target_table->chunk_count() == 0) {
       _target_table->append_mutable_chunk();
     }
 
-    start_chunk_id = _target_table->chunk_count() - 1;
-    end_chunk_id = start_chunk_id + 1;
-    auto last_chunk = _target_table->get_chunk(start_chunk_id);
-    start_index = last_chunk->size();
+    while (remaining_rows > 0) {
```

Comment:
An assumption all through the code base is that the last chunk is the one that will be inserted into. Here, we could end up with a full last chunk that might even later be compressed. Not sure about the implications, but maybe we want to avoid this?

Reply:
Don't see a problem, can you elaborate? Ending up with full Chunks is okay, is it not?

Reply:
Not sure if there is something in the code that would actually break right now. I am just wondering if we should always have a modifiable chunk at the end as this is how we communicate it, e.g., in our chunk visualization. Saying "the last chunk is where we insert" seems clearer that "the last chunk might be an encoded segment or a ValueSegment, and if its size is less than the maximum_chunk_size, this is where we insert".

Reply:
@Bouncner was involved in a couple of chunk layout things, too.

Reply:
Having the least amount of runtime-constraints like "you can rely on the last chunk being non-full and modifiable" seems desirable for me. Remember how well this "a table will always have at least one chunk"-thing went?